core/vm/contracts.go

```diff
@@ -498,3 +505,472 @@ func intToByteArray20(n uint8) [20]byte {
+	}
+	beneficiaryQi := common.BytesToAddress(input[:20], evm.chainConfig.Location)
+	value := new(big.Int).SetBytes(input[20:52])
+	etxGasLimit := binary.BigEndian.Uint64(input[52:60])
+
+	if *gas < etxGasLimit {
+		return ErrOutOfGas
+	}
+	*gas -= etxGasLimit
+
+	lockupContractAddress := LockupContractAddresses[[2]byte{evm.chainConfig.Location[0], evm.chainConfig.Location[1]}]
+	lockupContractAddressInternal, err := lockupContractAddress.InternalAndQuaiAddress()
+	if err != nil {
+		return err
+	}
+
+	_, err = beneficiaryQi.InternalAndQiAddress()
+	if err != nil {
+		return err
+	}
+	ownerContractInternal, err := ownerContract.InternalAndQuaiAddress()
+	if err != nil {
+		return err
+	}
+	ownerContractHash := common.BytesToHash(ownerContractInternal[:])
+
+	balanceHash := evm.StateDB.GetState(lockupContractAddressInternal, ownerContractHash)
+	if balanceHash == (common.Hash{}) {
+		return errors.New("no wrapped Qi balance in contract to unwrap")
+	}
+	balanceBig := balanceHash.Big()
+	if balanceBig.Cmp(value) == -1 {
+		return fmt.Errorf("not enough wrapped Qi to unwrap: have %v want %v", balanceBig, value)
+	}
+	balanceBig.Sub(balanceBig, value)
+	evm.StateDB.SetState(lockupContractAddressInternal, ownerContractHash, common.BigToHash(balanceBig))
+
+	evm.ETXCacheLock.RLock()
+	index := len(evm.ETXCache)
+	evm.ETXCacheLock.RUnlock()
+	if index > math.MaxUint16 {
+		return fmt.Errorf("CreateETX overflow error: too many ETXs in cache")
+	}
+
+	externalTx := types.ExternalTx{Value: value, To: &beneficiaryQi, Sender: ownerContract, EtxType: uint64(types.CoinbaseLockupType), OriginatingTxHash: evm.Hash, ETXIndex: uint16(index), Gas: etxGasLimit}
+	log.Global.Infof("Unwrapped Qi: %v -> %v, value: %v", ownerContract, beneficiaryQi, value)
+	evm.ETXCacheLock.Lock()
+	evm.ETXCache = append(evm.ETXCache, types.NewTx(&externalTx))
+	evm.ETXCacheLock.Unlock()
+	return nil
+}
+
+// WrapQi is called by the state processor to process an inbound Qi wrapping ETX
+// It stores the wrapped Qi balance in the lockup contract keyed with the contract address and provided Quai beneficiary address
+// To accept the deposit, the smart contract must call the ClaimQiDeposit function on the precompile
+func WrapQi(statedb StateDB, ownerContract, beneficiary common.Address, sender common.InternalAddress, value *big.Int, location common.Location) error {
+	ownerContractInternal, err := ownerContract.InternalAndQuaiAddress()
+	if err != nil {
+		return err
+	}
+	beneficiaryInternal, err := beneficiary.InternalAndQuaiAddress()
+	if err != nil {
+		return err
+	}
+	if sender != common.OneInternal(location) {
+		return errors.New("sender is not the correct internal address")
+	}
+	lockupContractAddress := LockupContractAddresses[[2]byte{location[0], location[1]}]
+	lockupContractAddressInternal, err := lockupContractAddress.InternalAndQuaiAddress()
+	if err != nil {
+		return err
+	}
+
+	if value.Sign() == -1 {
+		return errors.New("negative value")
+	}
+	if value.Sign() == 0 {
```

Comment:
if the value is zero should we reject this transaction?

Reply:
No, it's just a no-op